### PR TITLE
Determine namespace from file path

### PR DESCRIPTION
**Pull request by voronkovich**
Fix https://github.com/phpactor/phpactor/issues/558

### CODE REVIEW

**Comment by dantleech** [comment]
This looks perfect thanks.

Would it be possible to add a [unit test case](https://github.com/phpactor/phpactor/blob/develop/tests/Unit/Extension/SourceCodeFilesystem/Rpc/ClassSearchHandlerTest.php) for the handler and bonus points if you can add a simple [VIM regression test](https://github.com/phpactor/phpactor/blob/develop/plugin/tests/rpc_file_references.vader)?

**Comment by voronkovich** [comment]
@dantleech, I don't know what to test in Vim. I've added unit and system tests.

File: lib/Extension/ClassToFile/Application/FileInfo.php [hunk 1]
```diff
@@ -6,7 +6,7 @@
 use Phpactor\ClassFileConverter\Domain\FilePath;
 use Phpactor\ClassFileConverter\Domain\FileToClass;
 
-final class FileInfo
+class FileInfo
```
Comment (voronkovich) [inline lib/Extension/ClassToFile/Application/FileInfo.php]: Prophecy doesn't know how to test final classes.

File: tests/System/Extension/ClassToFile/Rpc/FileInfoTest.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Phpactor\Tests\System\Extension\ClassToFile\Rpc;
+
+use Phpactor\Tests\System\SystemTestCase;
+
+class FileInfoTest extends SystemTestCase
```
Comment (dantleech) [inline tests/System/Extension/ClassToFile/Rpc/FileInfoTest.php]: This one is not necessary I think, at least there is no precedent for RPC system tests at the moment, so reluctant to duplicate effort here.

The test I suggested was for the VIM plugin, which would have been an E2E test. These are few and are kept in `plugin/tests/*`, this one could look like:

```
# plugin/tests/get_namespace.vader
Execute:
    execute "edit src/Phpactor.php"
    namespace = call phpactor#GetNamespace()
    AssertEqual "Phpactor", namespace
```

Like I said, you would need the [vader](https://github.com/junegunn/vader.vim) plugin to execute this (with `:Vader` from within VIM with that file open).

But, if you don't want to do this, I can put it in later. No worries.

File: tests/Unit/Extension/ClassToFile/Rpc/FileInfoHandlerTest.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace Phpactor\Tests\Unit\Extension\ClassToFile\Rpc;
+
+use Phpactor\Extension\ClassToFile\Application\FileInfo;
+use Phpactor\Extension\ClassToFile\Rpc\FileInfoHandler;
+use Phpactor\Extension\Rpc\Response\ReturnResponse;
+use PHPUnit\Framework\TestCase;
+
+class FileInfoHandlerTest extends TestCase
```
Comment (dantleech) [inline tests/Unit/Extension/ClassToFile/Rpc/FileInfoHandlerTest.php]: This shoud extend `HandlerTestCase`, and would look something like:

```php
class FileInfoHandlerTest extends HandlerTestCase
{
    private $fileInfo;

    public function setUp()
    {
        $this->fileInfo = $this->prophesize(FileInfo::class);
    }

    protected function createHandler(): Handler
    {
        return new FileInfoHandler($this->fileInfo->reveal());
    }

    public function testReturnsAResponseWithAFileInfo()
    {
        $result = [
            'class' => 'App\Controller\BlogController',
            'class_name' => 'BlogController',
            'class_namespace' => 'App\Controller',
        ];

        $this->fileInfo->infoForFile($path)->willReturn($result);

        $response = $this->handle('file_info', [
            'path' => self::EXAMPLE_PATH
        ]);

        $this->assertInstanceOf(ReturnResponse::class, $response);
        $this->assertEquals($result, $response->parameters()['value']);
    }
}


```

File: tests/System/Extension/ClassToFile/Rpc/FileInfoTest.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        $response = json_decode($process->getOutput(), true);
+
+        $this->assertSuccess($process);
+        $this->assertEquals([
+            'action' => 'return',
+            'parameters' => [
+                'value' => [
+                    'class' => 'Animals\Badger\Carnivorous',
+                    'class_name' => 'Carnivorous',
+                    'class_namespace' => 'Animals\Badger',
```
Comment (dantleech) [inline tests/System/Extension/ClassToFile/Rpc/FileInfoTest.php]: Would it be worth adding `GetClassFullName` and `GetClassShortName` VIM functions while we are here?

**Comment by voronkovich** [comment]
@dantleech, I've fixed all issues you pointed out. Also I added the `GetClassFullName` method as you suggested. Such method could be helpful when it's needed to detect an alternate file (tests, form type and etc.). To get just a class name it's better to use `expand('%:t:r')`.

**Comment by dantleech** [comment]
Great, thanks for your efforts :) I'll add this to the CHANGELOG file.

**Pull request merged**